### PR TITLE
update dependencies

### PR DESCRIPTION
**Pull request by exterm**
- use Rails 7
- use Ruby 3.0 as default
- test against `ruby: [2.6, 2.7, 3.0]` and `rails: [6.0, main]`

## What are you trying to accomplish?
Packwerk is affected by the `mimemagic` dilemma like so many other projects, since the mimemagic version we were using was pulled. I used this as a trigger to update all our dependencies.

### Followup

- [ ] change branch protection rules to require new CI steps

### CODE REVIEW

File: test/unit/file_processor_test.rb [hunk 1]
```diff
@@ -46,20 +46,20 @@ class FileProcessorTest < Minitest::Test
...
+      @node_processor.expects(:call).with do |node, ancestors|
         parent = ancestors.first # class Hello; world; end
         Node.constant?(node) && # Hello
           Node.constant_name(node) == "Hello" &&
           ancestors.length == 1 &&
           Node.class?(parent) &&
           Node.class_or_module_name(parent) == "Hello"
       end
-      @node_processor.expects(:call).with do |node, ancestors:|
+      @node_processor.expects(:call).with do |node, ancestors|
```
Comment (exterm) [inline test/unit/file_processor_test.rb:62]: This threw a `ArgumentError: missing keyword: :ancestors` with Ruby 3.0. I assume there's a bug with mocha and Ruby 3, but I really don't want to go down that rabbit hole right now so I just changed the method to not use keyword arguments. I don't think it significantly degrades the design.

I added a sorbet signature to make up for it 😉

File: test/unit/cli_test.rb [hunk 2]
```diff
@@ -18,34 +18,26 @@ class CliTest < Minitest::Test
...
-    end
-
-    class FakeRunContext < Packwerk::RunContext
-      def initialize; end
-    end
-
-    class FakeConfiguration < Packwerk::Configuration
-      def initialize; end
-    end
-
```
Comment (exterm) [inline test/unit/cli_test.rb:32]: new style guide rules rightly rejected this; it doesn't initialize the objects correctly. The fix was a little more involved but I discovered that actually non of these fake classes are necessary.

**Comment by exterm** [comment]
This turned out to be quite the rabbit hole. I'd advise reviewers to go through commit by commit.

**Comment by exterm** [comment]
I am thinking about deconstructing this into multiple PRs.

**Comment by exterm** [comment]
I investigated but...

- Rails 7 needs at least Ruby 2.7
- I want to test against the two most recent major Ruby versions, so it is 3.0 and 2.7 then
- lots of gems are incompatible with Ruby 3.0 in their old versions, so the y need to be updated

everything else follows from that. So I think we should merge this PR as is.

**Comment by exterm** [comment]
minimal fix for just mimemagic lives in https://github.com/Shopify/packwerk/pull/111

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -8,8 +8,8 @@ jobs:
     strategy:
       matrix:
         ruby:
-          - 2.6
```
Comment (rafaelfranca) [inline .github/workflows/ci.yml:11]: We need to change the gemspec to require Ruby >= 2.7, but Ruby 2.6 is still supported at Shopify so I think we should continue to support here.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -29,7 +29,7 @@ jobs:
       - name: Set up Ruby 2.6
```
Comment (rafaelfranca) [inline .github/workflows/ci.yml:29]: This is now wrong

Comment (exterm) [reply .github/workflows/ci.yml:29]: ugh right - good catch. Not sure why the version needs to be in the name anyway

File: .github/workflows/ci.yml (hunk 3 above)
Comment (exterm) [reply .github/workflows/ci.yml:11]: That would mean that I need to test against two different Rails versions, as Rails 7 doesn't support Ruby 2.6.

**Comment by exterm** [comment]
Fixed everything @rafaelfranca @gmcgibbon 

The only thing I am not happy about is the size of the changeset, but I haven't managed to significantly reduce it for the reasons lined out in https://github.com/Shopify/packwerk/pull/110#issuecomment-811255851

File: .github/workflows/ci.yml (hunk 3 above)
Comment (exterm) [reply .github/workflows/ci.yml:11]: Fixed this by setting up a proper build matrix and multiple Gemfiles.

**Pull request merged**